### ISSUE TITLE
read-only mount fails on local disk

### ISSUE THREAD

**Issue by rodwalker** [hpc/charliecloud#401]
Hi,
I see the error
ch-run[38718]: can't re-mount image read-only (is it on NFS?): Operation not permitted (charliecloud.c:230 1)
but the image is on a local disk partition
/dev/sda7 on /tmp type ext3 (rw,nosuid,relatime,data=ordered)

I can live with the read-write option for now, but would be interested to know what feature of /tmp makes it look like NFS.

Cheers,
Rod.

**Comment by j-ogas** [comment]
Hello @rodwalker, that is a generic error message that occurs when the mount syscall,  
`mount(NULL, c->newroot, NULL, MS_REMOUNT|MS_BIND|MS_RDONLY, NULL)`, fails. This typically occurs when attempting to use ch-run on a NFS without the read-write option. Thus, we added the NFS hint. 

Related: #9, #7.

In this case, the remount as read-only on your disk partitioned `/tmp` is failing. I am in the process of building a vm with a manually partioned (non-tmpfs) tmp in an effort to reproduce.

**Comment by j-ogas** [comment]
I am unable to reproduce the error in a simple centos 7.6 virtualbox environment. 

```
$ df -Th | grep ^/dev/sda2
dev/sda2               ext3      1.9G  354M  1.4G  21% /tmp
$ ./bin/ch-run /tmp/centos7/ -- /bin/true
$ echo $?
0
```

There could be an issue with your /tmp partition.

**Comment by rodwalker** [comment]
Hi,
The machine is SLES12 SP3, so it would not surprise me if there is something odd. /var/tmp is ok!

rod/charliecloud/bin/ch-run  /tmp/centos7_atlas.0  pwd
ch-run[29392]: can't re-mount image read-only (is it on NFS?): Operation not permitted (charliecloud.c:230 1)

rod/charliecloud/bin/ch-run  /var/tmp/centos7_atlas.0  pwd
/
The respective partitions are
/dev/sda7 on /tmp type ext3 (rw,nosuid,relatime,data=ordered)
/dev/sda5 on /var type ext3 (rw,relatime,data=ordered)

so only difference nosuid. Unless the size is relevant for some reason
/dev/sda5       4,8G  885M  3,7G  19% /var
/dev/sda7       227G   69G  147G  32% /tmp

Cheers,
Rod.

**Comment by j-ogas** [comment]
Aha! 

That does seem to be the issue. I added `nosuid` to my /tmp partition and remounted it:

```
$ ./bin/ch-run /tmp/centos7/ -- /bin/true
ch-run[14920]: can't re-mount image read-only (is it on NFS?): Operation not permitted (charliecloud.c:230 1)
```

I bring this to attention of @reidpr. I suspect we may want to document this and/or update the error message. Thank you @rodwalker.

Best,
Jordan

**Comment by olifre** [comment]
That's a cool finding!
That also made me find this very related discussion:
https://github.com/moby/moby/issues/22994
It seems that the behaviour greatly varies for different kernel versions. However, what appears to work for most kernel versions (sadly, not all) seems to be to keep existing, not changeable options on the mount when remounting. So in this case, keeping `nosuid` when doing the read-only remount might help.

**Comment by j-ogas** [comment]
>  So in this case, keeping `nosuid` when doing the read-only remount might help.

FWIIW, adding `MS_NOSUID` to the mount flags did address this issue in my prescribed test environment, and it doesn't appear to have introduced regressions in the test suite.

**Comment by reidpr** [comment]
@rodwalker, this should be fixed in master now (or 0.9.9, which will be released shortly). When you have time, can you please confirm that this bug is fixed for you?
